### PR TITLE
[WIP] Add 3d convolution

### PR DESCRIPTION
**Pull request by ywatanabex**
I have implemented 3d convolution.
This enhancement is proposed in #945 .

Though the code passes tests on my environment. (with cudnn v4 and gpu),
I need to improve it to support various environment.

Beforehand, I would like to obtain some feedback.

### CODE REVIEW

**Comment by ywatanabex** [comment]
I would like to know the following points:

- whether the name (height, width, depth) is appropriate?  should call (dim1, dim2, dim3) or (x, y, z) ?
- what is the appropriate name of 3-d version of im2col?
- Is it good to use `workspace_limit` option?

**Comment by jheymann85** [comment]
Regarding the second point: I think it's called vol2col and col2vol in the caffee implementation which sounds appropriate in my point of view.

**Comment by delta2323** [comment]
Thank you for your PR! We will check the implementation. I have several comments.

cuDNN has primitives that are necessary for N-dimensional convolution. What is the main difficulty for extending convolution to arbitrary dimensions? (I do not have clear answer)

> whether the name (height, width, depth) is appropriate? should call (dim1, dim2, dim3) or (x, y, z) ?
Do you use `(height, width, depth)` in the implementation (I cannot found it)?

> I think it's called vol2col and col2vol in the caffee implementation which sounds appropriate in my point of view.
Are you thinking of [C3D](https://github.com/facebook/C3D) by facebook, which is modified version of Caffe. I agree with your suggestion.

**Comment by delta2323** [comment]
I would like to use this implementation for [deepmark](https://github.com/DeepMark/deepmark).

File: chainer/functions/connection/convolution_3d.py [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+            libcudnn.CUDNN_CONVOLUTION_BWD_DATA_SPECIFY_WORKSPACE_LIMIT
+
+
+def _triplet(x):
+    if hasattr(x, '__getitem__'):
+        return x
+    return (x, x, x)
+
+
+def im32col_cpu(img, kh, kw, kd, sy, sx, sz, ph, pw, pd, pval=0, cover_all=False):
```
Comment (delta2323) [inline chainer/functions/connection/convolution_3d.py]: `im2col_[cg]pu` and `col2im_[cg]pu` are implemented in `utils/conv`. So I think it is appropriate to implement 3d-version there, too.

File: chainer/functions/connection/convolution_3d.py [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+            dtype = x.dtype
+            one = numpy.array(1, dtype=dtype).ctypes
+            zero = numpy.array(0, dtype=dtype).ctypes
+            libcudnn.convolutionForward(
+                handle, one.data, x_desc.value, x.data.ptr,
+                self.filter_desc.value, W.data.ptr, self.conv_desc.value,
+                algo, workspace.data.ptr, workspace_size, zero.data,
+                y_desc.value, y.data.ptr)
+
+            # TODO(beam2d): Support unshared bias
```
Comment (delta2323) [inline chainer/functions/connection/convolution_3d.py]: Fix the author of TODO.

File: chainer/functions/connection/convolution_3d.py [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+
+def convolution_3d(x, W, b=None, stride=1, pad=0, use_cudnn=True):
+    """Three-dimensional convolution function.
+    """
+    func = Convolution3DFunction(stride, pad, use_cudnn)
+    if b is None:
+        return func(x, W)
+    else:
+        return func(x, W, b)
+
```
Comment (delta2323) [inline chainer/functions/connection/convolution_3d.py]: Remove trailing blank lines.

File: chainer/functions/connection/convolution_3d.py [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+                y_desc.value, y.data.ptr)
+
+            # TODO(beam2d): Support unshared bias
+            if b is not None:
+                libcudnn.addTensor(
+                    handle, one.data,
+                    self.bias_desc.value, b.data.ptr, one.data,
+                    y_desc.value, y.data.ptr)
+        else:
+            raise NotImplementedError('forward_gpu without cudnn is not implemented yet.')
```
Comment (delta2323) [inline chainer/functions/connection/convolution_3d.py]: cuDNN (capitalize DNN)

File: chainer/functions/connection/convolution_3d.py [hunk 5]
```diff
@@ -0,0 +1,279 @@
...
+                    gy_desc.value, gy.data.ptr, self.conv_desc.value,
+                    zero.data, x_desc.value, gx.data.ptr)
+
+            if b is not None:
+                gb = cuda.cupy.empty_like(b)
+                libcudnn.convolutionBackwardBias(
+                    handle, one.data, gy_desc.value, gy.data.ptr,
+                    zero.data, self.bias_desc.value, gb.data.ptr)
+        else:
+            raise NotImplementedError('forward_gpu without cudnn is not implemented yet.')
```
Comment (delta2323) [inline chainer/functions/connection/convolution_3d.py]: cuDNN (capitalize DNN)

File: tests/chainer_tests/functions_tests/connection_tests/test_convolution_3d.py [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+from chainer.testing import condition
+from chainer.utils import conv
+
+def _triplet(x):
+    if hasattr(x, '__getitem__'):
+        return x
+    return x, x, x
+
+
+def _asfortranarray(x):
```
Comment (delta2323) [inline tests/chainer_tests/functions_tests/connection_tests/test_convolution_3d.py]: asfortanarray is already implemented in both `numpy` and `cupy`. So we do not have to implement `asfortranarray` by yourself.

**Comment by delta2323** [comment]
Please add test cases which raises `NotImplementedError` when `use_cudnn` is `False`.

File: cupy/cuda/cudnn.pxd [hunk 7]
```diff
@@ -133,7 +133,9 @@ cpdef destroyTensorDescriptor(size_t tensorDesc)
 cpdef addTensor_v2(
         size_t handle, int mode, size_t alpha, size_t biasDesc,
         size_t biasData, size_t beta, size_t srcDestDesc, size_t srcDestData)
-
+cpdef addTensor(
```
Comment (delta2323) [inline cupy/cuda/cudnn.pxd]: The API of `addTensor` differs depending on the version of cuDNN. So use `addTensor_v2` or `addTensor_v3`. This fix is already done in the current master branch. So please merge or rebase the master branch to resolve this problem.

Comment (delta2323) [inline cupy/cuda/cudnn.pxd]: We had prepared `addTensor_v3` in the master branch (it is actually identical to `addTensor` if cuDNN is newer then v2). So please merge or rebase the master branch and use this API.

**Comment by delta2323** [comment]
If you are afford to do it, I would like to add `deterministic` option as is done in #1189. It is not a must because we can do it after this PR is merged.

File: chainer/functions/connection/convolution_3d.py [hunk 8]
```diff
@@ -0,0 +1,279 @@
...
+            zero = numpy.array(0, dtype=dtype).ctypes
+            libcudnn.convolutionForward(
+                handle, one.data, x_desc.value, x.data.ptr,
+                self.filter_desc.value, W.data.ptr, self.conv_desc.value,
+                algo, workspace.data.ptr, workspace_size, zero.data,
+                y_desc.value, y.data.ptr)
+
+            # TODO(beam2d): Support unshared bias
+            if b is not None:
+                libcudnn.addTensor(
```
Comment (delta2323) [inline chainer/functions/connection/convolution_3d.py]: Use `addTensor_v3`.

File: chainer/functions/connection/convolution_3d.py [hunk 9]
```diff
@@ -0,0 +1,279 @@
...
+
+        if b is None:
+            return gx, gW
+        else:
+            return gx, gW, gb
+
+
+def convolution_3d(x, W, b=None, stride=1, pad=0, use_cudnn=True):
+    """Three-dimensional convolution function.
+    """
```
Comment (delta2323) [inline chainer/functions/connection/convolution_3d.py]: Add the description of the arguments.

File: chainer/links/connection/convolution_3d.py [hunk 10]
```diff
@@ -0,0 +1,53 @@
+import numpy
+
+from chainer.functions.connection import convolution_3d
+from chainer import link
+
+
+class Convolution3D(link.Link):
+
+    """Three-dimensional convolutional layer.
+    """
```
Comment (delta2323) [inline chainer/links/connection/convolution_3d.py]: Please add document.

File: chainer/links/connection/convolution_3d.py [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+                 initialW=None, initial_bias=None):
+        kh, kw, kd = _triplet(ksize)
+        self.stride = _triplet(stride)
+        self.pad = _triplet(pad)
+        self.use_cudnn = use_cudnn
+
+        W_shape = (out_channels, in_channels, kh, kw, kd)
+        super(Convolution3D, self).__init__(W=W_shape)
+
+        if initialW is not None:
```
Comment (delta2323) [inline chainer/links/connection/convolution_3d.py]: Please use initializers to initialize the weight matrix and the bias vector.
I think the implementation of `Convolution2D` will be helpful.
https://github.com/pfnet/chainer/blob/master/chainer/links/connection/convolution_2d.py#L60

**Comment by delta2323** [comment]
As we would like to have more generalized (i.e. N-dimensional) convolution, we are now implementing it as PR #1258. Thank you for your cooperation and sorry for our late response.

**Comment by ywatanabex** [comment]
Hi, I really sorry for the delayed response and appreciate for the code review.
It's good to implement n-dim convolution. Actually, it's nothing difficult with the help of cuDNN.
I can help to implement it in some part.

**Comment by delta2323** [comment]
@ywatanabex Thank you for your response. Your original implementation greatly helps us to tackle ND Convolution. And I also appreciate you that you help us implementing it. It will be done in #1258.

**Comment by delta2323** [comment]
#1279 implements ConvolutionND and is now merged. So I will close this PR. Your contribution greatly helps to implements the PR. Thank you!

**Pull request closed without merging**